daemon/sync.go

```diff
@@ -0,0 +1,484 @@
+package daemon
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"strings"
+	"time"
+
+	"github.com/weaveworks/flux"
+	"github.com/weaveworks/flux/cluster"
+	"github.com/weaveworks/flux/event"
+	"github.com/weaveworks/flux/git"
+	fluxmetrics "github.com/weaveworks/flux/metrics"
+	"github.com/weaveworks/flux/resource"
+	fluxsync "github.com/weaveworks/flux/sync"
+	"github.com/weaveworks/flux/update"
+)
+
+// Sync holds the data we are working with during a sync.
+type Sync struct {
+	logger      log.Logger
+	working     *git.Checkout
+	oldTagRev   string
+	newTagRev   string
+	initialSync bool
+	commits     []git.Commit
+
+	*Daemon
```

Comment:
Including the Daemon here is convenient -- but brittle, on a couple of counts:

 - because it's typed as *Daemon, the code accessing it could depend on anything in that struct
 - because it's embedded, it is hard to see in the code which method or field access are dependencies

I think providing the *git.Working to the entry point (either as an argument to `Run` or when initialising a `struct Sync`) would obviate much of the need for `*Daemon`. Going through other implicit uses of *Daemon I can find:

 - `GitConfig`, or the relevant fields, can be supplied up front too
 - The syncSetName can be calculated up front and supplied
 - `lastKnownSyncTagRev`/`warnAboutSyncTagChange` -- either encode the required interactions into an interface, and supply it at the start, or take the values (as a struct) in an arg, and return the new values at the end.

No doubt I've missed some uses and underestimated the complexity. If these are just too difficult to disentangle, another way to advance would be to replace the `*Daemon` field with an interface containing only the dependencies.

Reply:
Took most of your comments into account and succeeded (with some pain along the way) to turn everything into stateless functions.

I reckon the `Run()` method needs some in-code docs / annotations to help someone navigate through the function calls, and some log and error messages need a more descriptive message.